### ISSUE TITLE
BLE Listener and Tesla Control Issues with M5nanoC6 Dev Kit on Apple Silicon, macOS

### ISSUE THREAD

**Issue by danielraffel** [yoziru/esphome-tesla-ble#93]
### TL;DR

I’m trying to install and use `esphome-tesla-ble` on an M5nanoC6 dev kit using an Apple Silicon MacBook pro running macOS Sequioa. While I’ve made progress installing and tweaking the repo, I’m encountering compilation warnings and issues related to BLE functionality. Despite successfully uploading the firmware and seeing the device in Home Assistant, the listener doesn’t seem to work as expected, and I’m unable to control Tesla features like charge limit or amperage. I’m unsure whether the issue lies with BLE-specific compilation errors or something else.

---

### Steps Taken
_Initial steps: I installed the required [macOS drivers](https://github.com/WCHSoftGroup/ch34xser_macos?tab=readme-ov-file) selecting the DMG option since I am on Apple Silicon, restarted the system, and enabled the driver by navigating to `System Settings > General > Login Items & Extensions > Extensions`, selecting the info icon for `Driver Extensions`, and enabling the `CH34xVCPDriver`. After that, I opened a terminal and ran `python3 -m pip install esphome`. Then, I cloned the `esphome-tesla-ble` repo and..._

1. **Configured `secrets.yaml`**:
   Added the following:
   ```yaml
   wifi_ssid: my_wifi
   wifi_password: my_password
   wifi_hotspot_password: my_hotspot_password
   api_encryption_key: my_encryption_key
   tesla_vin: my_tesla_vin
   ```

2. **Modified `boards/esp32-generic.yml`**:
   Adjusted for the M5nanoC6 dev kit:
   ```yaml
   substitutions:
     board: esp32-c6-devkitc-1
     variant: esp32c6
     flash_size: "4MB"

   esp32:
     board: ${board}
     framework:
       type: esp-idf
   ```

3. **Enabled Listener in `tesla-ble-esp32-generic.yml`**:
   Uncommented:
   ```yaml
   listener: !include packages/listener.yml # Uncomment this to scan find your VIN BLE MAC address
   ```

4. **Compiled with ESPHome**:
   Ran `esphome compile tesla-ble-esp32-generic.yml`, successfully compiled but resulting in the following warning:

```
Compiling .pioenvs/tesla-ble/bt/common/osi/fixed_queue.c.o
In file included from src/esphome.h:35,
                 from src/main.cpp:3:
src/esphome/components/esp32_ble/const_esp32c6.h:69:1: warning: missing initializer for member 'esp_bt_controller_config_t::cca_drop_mode' [-Wmissing-field-initializers]
   69 | };
      | ^
src/esphome/components/esp32_ble/const_esp32c6.h:69:1: warning: missing initializer for member 'esp_bt_controller_config_t::cca_low_tx_pwr' [-Wmissing-field-initializers]
Compiling .pioenvs/tesla-ble/bt/common/osi/pkt_queue.c.o
```

5. **Uploaded Firmware**:
   Used `esphome upload tesla-ble-esp32-generic.yml` and when prompted selected to uploading using `/dev/cu.usbmodem1101 (USB JTAG/serial debug unit)`, and Successfully uploaded program.

6. **Checked Logs in Home Assistant**:
When I unplug and restart the esp chip I see it in Home Assistant. If I navigate to `settings > devices > ESPHome` and `Enable Debug logging` then switch over to `settings > systems > logs` and select the overflow menu (three dots) `show raw logs` the logs didn’t match the expected listener output (per [documentation](https://github.com/yoziru/esphome-tesla-ble?tab=readme-ov-file#finding-the-ble-mac-address-of-your-vehicle)) I just see stuff like this:

```
2024-11-24 22:27:42.502 DEBUG (MainThread) [aioesphomeapi._frame_helper.base] tesla-ble-5ae9cc @ 192.168.86.55: Sending frame: [0100149659e1d330d6179c9647f7c4e869f7fc2869935c]
2024-11-24 22:27:42.568 DEBUG (MainThread) [aioesphomeapi.connection] tesla-ble-5ae9cc @ 192.168.86.55: Got message of type PingRequest: 
2024-11-24 22:27:42.568 DEBUG (MainThread) [aioesphomeapi.connection] tesla-ble-5ae9cc @ 192.168.86.55: Sending PingResponse: 
2024-11-24 22:27:42.569 DEBUG (MainThread) [aioesphomeapi._frame_helper.base] tesla-ble-5ae9cc @ 192.168.86.55: Sending frame: [01001470464aa79669e1baf2cf3d402c85f31c527d0ff6]
2024-11-24 22:27:42.571 DEBUG (MainThread) [aioesphomeapi.connection] tesla-ble-5ae9cc @ 192.168.86.55: Got message of type PingResponse: 
2024-11-24 22:27:51.350 DEBUG (MainThread) [aioesphomeapi.connection] tesla-ble-5ae9cc @ 192.168.86.55: Got message of type SensorStateResponse: key: 1324261225
state: 217.435
2024-11-24 22:28:19.343 DEBUG (MainThread) [aioesphomeapi.connection] tesla-ble-5ae9cc @ 192.168.86.55: Got message of type SensorStateResponse: key: 799351157
state: -68
2024-11-24 22:28:42.505 DEBUG (MainThread) [aioesphomeapi.connection] tesla-ble-5ae9cc @ 192.168.86.55: Sending PingRequest: 
2024-11-24 22:28:42.505 DEBUG (MainThread) [aioesphomeapi._frame_helper.base] tesla-ble-5ae9cc @ 192.168.86.55: Sending frame: [010014322542678ff2db170e661aafca260edaaa4c341e]
2024-11-24 22:28:42.517 DEBUG (MainThread) [aioesphomeapi.connection] tesla-ble-5ae9cc @ 192.168.86.55: Got message of type PingResponse: 
2024-11-24 22:28:51.356 DEBUG (MainThread) [aioesphomeapi.connection] tesla-ble-5ae9cc @ 192.168.86.55: Got message of type SensorStateResponse: key: 1324261225
state: 277.435
```
---

### Curious Observation

I attempted to pair my Tesla’s Bluetooth with macOS, when I did I got a popup showing a BLE MAC ID. I assumed this might be the MAC ID needed for `secrets.yaml`, but after adding it, I still couldn’t control features like amperage or charge limit.

Here’s a screenshot of the pairing popup:  
![image](https://github.com/user-attachments/assets/3f94e546-6de9-4596-8bbc-5555bdc2b43e)
---

### Expectations vs. Observations

1. **Compilation**:
   - **Expectation**: No warnings or missing initializers during compilation.
   - **Observation**: Received warnings related to uninitialized fields `cca_drop_mode` and `cca_low_tx_pwr` in `const_esp32c6.h`. These might indicate compatibility issues with the ESP-IDF framework or BLE-specific functionality.

2. **Listener Functionality**:
   - **Expectation**: Logs should show BLE listener output similar to the [example](https://github.com/yoziru/esphome-tesla-ble?tab=readme-ov-file#finding-the-ble-mac-address-of-your-vehicle).
   - **Observation**: Logs only show `SensorStateResponse` messages, not the expected BLE discovery output.

3. **Tesla BLE Control**:
   - **Expectation**: Adding the BLE MAC ID I was shown when pairing macOS to `secrets.yaml` should allow me to control Tesla features like charge limit and amperage.
   - **Observation**: No control over Tesla features in the Home Assistant dashboard.

---

### Questions / Comments

1. Could the compilation warnings (`cca_drop_mode` and `cca_low_tx_pwr`) impact BLE functionality? How should these fields be initialized?

3. Why doesn’t the listener produce expected output in the logs? Could this be related to BLE-specific compilation issues or something else?

4. Is the BLE MAC ID from the macOS pairing popup valid for use in `secrets.yaml`, or should it be obtained differently? Could there be multiple BLE Tesla MAC IDs? This one started with: `4C:FC:AA`

---

### Current Assumptions

- The root issue may be related to BLE-specific compilation problems or an incomplete setup of the listener component.
- Resolving the `esp_bt_controller_config_t` initializer warnings might address these issues.

**I’d appreciate any guidance or troubleshooting steps to resolve these problems!**

**Comment by danielraffel** [comment]
Figured out my issues. Didn't realize I needed to use my key card twice. Wrote up instructions.

https://danielraffel.me/til/2024/11/27/how-i-set-up-an-esp32-ble-key-with-my-tesla-and-home-assistant-for-solar-charging-optimization/
